# Outage Severity

**Name(s)**: Kyle Shiroma, Marcus De Ramos

**Website Link**: (your website link)

In [236]:
import pandas as pd
import numpy as np
from pathlib import Path

import plotly.express as px
pd.options.plotting.backend = 'plotly'

# from dsc80_utils import * # Feel free to uncomment and use this.

## Step 1: Introduction

In [237]:
# Upload data and skip first 5 rows since its just info about the dataset
outage = (pd.read_excel('/Users/marcusderamos/Desktop/DSC 80 final proj/Power-Outages/outage.xlsx.xls',skiprows = 5) 
          .drop(index=0) # Drop  'units' row; was used to describe column values
          .reset_index(drop=True) # Reset index
          .drop(columns=['variables'])) # Drop 'variables' column; was used to descibe data structure
outage

#  Question:
# What makes some power outages worse than others — and can we predict that?


# U.S. Power Outage Dataset 

# COLUMN NAMES AND DEFINITION:

# OBS: Observation ID (unique outage record)
# YEAR: Year of outage
# MONTH: Month of outage (1–12)
# U.S._STATE: State where outage occurred
# POSTAL.CODE: State abbreviation

# NERC.REGION: Power grid reliability region North American Electric Reliability Corporation (NERC) 
    # WECC - Western U.S. (California, West Coast, Mountain states)
    # MRO - Midwest Reliability Organization (Upper Midwest)
    # SERC - Southeastern Electric Reliability Council (Southeast)
    # RFC - ReliabilityFirst Corporation (Mid-Atlantic / Ohio Valley)
    # NPCC - Northeast Power Coordinating Council (Northeast U.S.)
    # TRE - Texas Reliability Entity (Texas grid, largely independent)
    # FRCC - Florida Reliability Coordinating Council (Florida) 

# CLIMATE.REGION: U.S. climate zone classification
    # Northeast
    # East North Central
    # Central
    # Southeast
    # South
    # West North Central
    # Northwest
    # Southwest
    # West

# ANOMALY.LEVEL: How much hotter or colder it was than usual (in °C)
# CLIMATE.CATEGORY: Climate condition (normal / warm / cold)

# Outage Timing
# OUTAGE.START.DATE: Date outage began
# OUTAGE.START.TIME: Time outage began
# OUTAGE.RESTORATION.DATE: Date power restored
# OUTAGE.RESTORATION.TIM: Time power restored
# OUTAGE.DURATION: Duration of outage (minutes)

# Cause of Outage
# CAUSE.CATEGORY: General cause (weather, equipment, human error, etc.)
# CAUSE.CATEGORY.DETAIL: Specific cause description
# HURRICANE.NAMES: Hurricane name if applicable

# Impact Severity
# DEMAND.LOSS.MW: Electricity demand lost (megawatts)
# CUSTOMERS.AFFECTED: Number of customers without power

# Electricity Prices
# RES.PRICE: Residential electricity price
# COM.PRICE: Commercial electricity price
# IND.PRICE: Industrial electricity price
# TOTAL.PRICE: Overall average electricity price

# Electricity Sales
# RES.SALES: Residential electricity sales
# COM.SALES: Commercial sales
# IND.SALES: Industrial sales
# TOTAL.SALES: Total electricity sales

# Sales Percentages
# RES.PERCEN: % residential electricity usage
# COM.PERCEN: % commercial usage
# IND.PERCEN: % industrial usage

# Customer Counts
# RES.CUSTOMERS: Residential customers
# COM.CUSTOMERS: Commercial customers
# IND.CUSTOMERS: Industrial customers
# TOTAL.CUSTOMERS: Total customers

# Customer Percentages
# RES.CUST.PCT: % residential customers
# COM.CUST.PCT: % commercial customers
# IND.CUST.PCT: % industrial customers

# Economics
# PC.REALGSP.STATE: State per-capita real GDP
# PC.REALGSP.USA: U.S. per-capita real GDP
# PC.REALGSP.REL: State GDP relative to U.S.
# PC.REALGSP.CHANGE: Change in state GDP
# UTIL.REALGSP: Utility sector GDP
# TOTAL.REALGSP: Total state GDP
# UTIL.CONTRI: Utility sector contribution share
# PI.UTIL.OFUSA: Utility income % of U.S. total

# Population & Urbanization
# POPULATION: State population
# POPPCT_URBAN: % population in urban areas
# POPPCT_UC: % population in urban clusters (small towns/ suburbs)

# Population Density
# POPDEN_URBAN: Urban population density
# POPDEN_UC: Urban cluster density
# POPDEN_RURAL: Rural population density

# Land & Water Geography
# AREAPCT_URBAN: % land classified as urban
# AREAPCT_UC: % land urban clusters
# PCT_LAND: % of state area that is land
# PCT_WATER_TOT: % total water area
# PCT_WATER_INLAND: % inland water area


,OBS,YEAR,MONTH,U.S._STATE,POSTAL.CODE,NERC.REGION,CLIMATE.REGION,ANOMALY.LEVEL,CLIMATE.CATEGORY,OUTAGE.START.DATE,...,POPPCT_URBAN,POPPCT_UC,POPDEN_URBAN,POPDEN_UC,POPDEN_RURAL,AREAPCT_URBAN,AREAPCT_UC,PCT_LAND,PCT_WATER_TOT,PCT_WATER_INLAND
0,1.0,2011.0,7.0,Minnesota,MN,MRO,East North Central,-0.3,normal,2011-07-01 00:00:00,...,73.27,15.28,2279,1700.5,18.2,2.14,0.6,91.592666,8.407334,5.478743
1,2.0,2014.0,5.0,Minnesota,MN,MRO,East North Central,-0.1,normal,2014-05-11 00:00:00,...,73.27,15.28,2279,1700.5,18.2,2.14,0.6,91.592666,8.407334,5.478743
2,3.0,2010.0,10.0,Minnesota,MN,MRO,East North Central,-1.5,cold,2010-10-26 00:00:00,...,73.27,15.28,2279,1700.5,18.2,2.14,0.6,91.592666,8.407334,5.478743
3,4.0,2012.0,6.0,Minnesota,MN,MRO,East North Central,-0.1,normal,2012-06-19 00:00:00,...,73.27,15.28,2279,1700.5,18.2,2.14,0.6,91.592666,8.407334,5.478743
4,5.0,2015.0,7.0,Minnesota,MN,MRO,East North Central,1.2,warm,2015-07-18 00:00:00,...,73.27,15.28,2279,1700.5,18.2,2.14,0.6,91.592666,8.407334,5.478743
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1529,1530.0,2011.0,12.0,North Dakota,ND,MRO,West North Central,-0.9,cold,2011-12-06 00:00:00,...,59.9,19.9,2192.2,1868.2,3.9,0.27,0.1,97.599649,2.401765,2.401765
1530,1531.0,2006.0,NaN,North Dakota,ND,MRO,West North Central,NaN,NaN,NaN,...,59.9,19.9,2192.2,1868.2,3.9,0.27,0.1,97.599649,2.401765,2.401765
1531,1532.0,2009.0,8.0,South Dakota,SD,RFC,West North Central,0.5,warm,2009-08-29 00:00:00,...,56.65,26.73,2038.3,1905.4,4.7,0.3,0.15,98.307744,1.692256,1.692256
1532,1533.0,2009.0,8.0,South Dakota,SD,MRO,West North Central,0.5,warm,2009-08-29 00:00:00,...,56.65,26.73,2038.3,1905.4,4.7,0.3,0.15,98.307744,1.692256,1.692256


The data is from Purdue university in the form of an excel file. The goal of this project is to do an EDA of the dataset to figure out the severity of outages in a given area. More specifically, what characteristics are associated with higher severity major power outages? This will be steps 1-4 with data cleaning and eda, step 3 is assesment of missningness, and step 4 is hypothesis testing. 5-8 will be modeling based. Framing a prediction problem, baseline model, final model, and fairness analysis. 

## Step 2: Data Cleaning and Exploratory Data Analysis

In [238]:
# Create copy of data
outage_clean = outage.copy()

outage_clean

,OBS,YEAR,MONTH,U.S._STATE,POSTAL.CODE,NERC.REGION,CLIMATE.REGION,ANOMALY.LEVEL,CLIMATE.CATEGORY,OUTAGE.START.DATE,...,POPPCT_URBAN,POPPCT_UC,POPDEN_URBAN,POPDEN_UC,POPDEN_RURAL,AREAPCT_URBAN,AREAPCT_UC,PCT_LAND,PCT_WATER_TOT,PCT_WATER_INLAND
0,1.0,2011.0,7.0,Minnesota,MN,MRO,East North Central,-0.3,normal,2011-07-01 00:00:00,...,73.27,15.28,2279,1700.5,18.2,2.14,0.6,91.592666,8.407334,5.478743
1,2.0,2014.0,5.0,Minnesota,MN,MRO,East North Central,-0.1,normal,2014-05-11 00:00:00,...,73.27,15.28,2279,1700.5,18.2,2.14,0.6,91.592666,8.407334,5.478743
2,3.0,2010.0,10.0,Minnesota,MN,MRO,East North Central,-1.5,cold,2010-10-26 00:00:00,...,73.27,15.28,2279,1700.5,18.2,2.14,0.6,91.592666,8.407334,5.478743
3,4.0,2012.0,6.0,Minnesota,MN,MRO,East North Central,-0.1,normal,2012-06-19 00:00:00,...,73.27,15.28,2279,1700.5,18.2,2.14,0.6,91.592666,8.407334,5.478743
4,5.0,2015.0,7.0,Minnesota,MN,MRO,East North Central,1.2,warm,2015-07-18 00:00:00,...,73.27,15.28,2279,1700.5,18.2,2.14,0.6,91.592666,8.407334,5.478743
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1529,1530.0,2011.0,12.0,North Dakota,ND,MRO,West North Central,-0.9,cold,2011-12-06 00:00:00,...,59.9,19.9,2192.2,1868.2,3.9,0.27,0.1,97.599649,2.401765,2.401765
1530,1531.0,2006.0,NaN,North Dakota,ND,MRO,West North Central,NaN,NaN,NaN,...,59.9,19.9,2192.2,1868.2,3.9,0.27,0.1,97.599649,2.401765,2.401765
1531,1532.0,2009.0,8.0,South Dakota,SD,RFC,West North Central,0.5,warm,2009-08-29 00:00:00,...,56.65,26.73,2038.3,1905.4,4.7,0.3,0.15,98.307744,1.692256,1.692256
1532,1533.0,2009.0,8.0,South Dakota,SD,MRO,West North Central,0.5,warm,2009-08-29 00:00:00,...,56.65,26.73,2038.3,1905.4,4.7,0.3,0.15,98.307744,1.692256,1.692256


In [239]:
# Drop columns we don't need
drop_cols = ['POSTAL.CODE','CLIMATE.REGION','CAUSE.CATEGORY.DETAIL','HURRICANE.NAMES',
             'RES.PRICE','COM.PRICE','IND.PRICE','RES.SALES','COM.SALES','IND.SALES',
             'TOTAL.SALES','RES.PERCEN','IND.PERCEN','RES.CUSTOMERS','COM.CUSTOMERS',
             'IND.CUSTOMERS','RES.CUST.PCT','COM.CUST.PCT','IND.CUST.PCT','PC.REALGSP.USA',
             'PC.REALGSP.REL','PC.REALGSP.CHANGE','UTIL.REALGSP','TOTAL.REALGSP','UTIL.CONTRI',
             'PI.UTIL.OFUSA','POPPCT_UC','POPDEN_UC','AREAPCT_URBAN','AREAPCT_UC','PCT_LAND',
             'PCT_WATER_TOT','PCT_WATER_INLAND']

outage_clean = outage_clean.drop(columns=drop_cols)


In [240]:
# List of columns that need to be converted to numbers
numeric_cols = ['OUTAGE.DURATION','CUSTOMERS.AFFECTED','DEMAND.LOSS.MW',
                'YEAR','MONTH','ANOMALY.LEVEL','COM.PERCEN','TOTAL.PRICE',
                'TOTAL.CUSTOMERS','PC.REALGSP.STATE','POPULATION',
                'POPPCT_URBAN','POPDEN_URBAN','POPDEN_RURAL']

# Converting numeric_cols to numbers
for col in numeric_cols: outage_clean[col] = pd.to_numeric(outage_clean[col], errors='coerce')  # coerce replaces bad values with NaN

In [241]:
# Combine 'OUTAGE.START.DATE' & 'OUTAGE.START.TIME' into one column and name it 'OUTAGE.START'
outage_clean['OUTAGE.START'] = pd.to_datetime(outage_clean['OUTAGE.START.DATE'].astype(str) 
                                        + ' ' 
                                        + outage_clean['OUTAGE.START.TIME'].astype(str),errors='coerce')

# Combine 'OUTAGE.RESTORATION.DATE' & 'OUTAGE.RESTORATION.TIME' and name it 'OUTAGE.RESTORATION'
outage_clean['OUTAGE.RESTORATION'] = pd.to_datetime(outage_clean['OUTAGE.RESTORATION.DATE'].astype(str) 
                                              + ' ' 
                                              + outage_clean['OUTAGE.RESTORATION.TIME'].astype(str),
                                              errors='coerce')

# Drop the original date/time columns since we combined them
outage_clean = outage_clean.drop(columns=['OUTAGE.START.DATE','OUTAGE.START.TIME','OUTAGE.RESTORATION.DATE','OUTAGE.RESTORATION.TIME'])

/var/folders/1b/54c6c0t12kn6cl16xckz50w80000gn/T/ipykernel_4593/4028737641.py:2: UserWarning:

Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.

/var/folders/1b/54c6c0t12kn6cl16xckz50w80000gn/T/ipykernel_4593/4028737641.py:7: UserWarning:

Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.



In [242]:
# Make column 'YEAR' and 'MONTH' into ints
outage_clean['YEAR'] = outage_clean['YEAR'].astype('Int64')
outage_clean['MONTH'] = outage_clean['MONTH'].astype('Int64')

In [243]:
# Remove Alaska because we want to focus on the mainland
outage_clean = outage_clean[outage_clean['U.S._STATE'] != 'Alaska']

In [244]:
# Creating new columns using FEATURE ENGINEERING

# Density Gap: shows how much denser cities are compared to rural areas
outage_clean['DENSITY_GAP'] = outage_clean['POPDEN_URBAN'] / outage_clean['POPDEN_RURAL'].replace(0, np.nan)


# Create 'DENSITY_TIER' column: Split U.S. states into 3 tiers based on density gap
    # qcut(q=3) splits the data into thirds
    # Low Gap: Population is more evenly spread out
    # Medium Gap: Balanced urban–rural difference
    # High Gap: Large difference between city and rural population density
outage_clean['DENSITY_TIER'] = pd.qcut(outage_clean['DENSITY_GAP'].dropna(), q=3, labels=['Low Gap', 'Medium Gap', 'High Gap'])

# MISSINGNESS
# Create column showing if CUSTOMERS.AFFECTED is missing
    # True: missing value
    # False: has value
outage_clean['CUST_MISSING'] = outage_clean['CUSTOMERS.AFFECTED'].isna()

# Create column showing if outage is severe (top 25% longest outages)
outage_clean['SEVERE'] = (outage_clean['OUTAGE.DURATION'] >= outage_clean['OUTAGE.DURATION'].quantile(0.75)).astype('Int64')

# Check dataset size and preview results
print(outage_clean.shape)
print('POPDEN_URBAN' in outage_clean.columns)
outage_clean.head()

(1533, 25)
True


,OBS,YEAR,MONTH,U.S._STATE,NERC.REGION,ANOMALY.LEVEL,CLIMATE.CATEGORY,CAUSE.CATEGORY,OUTAGE.DURATION,DEMAND.LOSS.MW,...,POPULATION,POPPCT_URBAN,POPDEN_URBAN,POPDEN_RURAL,OUTAGE.START,OUTAGE.RESTORATION,DENSITY_GAP,DENSITY_TIER,CUST_MISSING,SEVERE
0,1.0,2011,7,Minnesota,MRO,-0.3,normal,severe weather,3060.0,NaN,...,5348119.0,73.27,2279.0,18.2,2011-07-01 17:00:00,2011-07-03 20:00:00,125.21978,Medium Gap,False,1
1,2.0,2014,5,Minnesota,MRO,-0.1,normal,intentional attack,1.0,NaN,...,5457125.0,73.27,2279.0,18.2,2014-05-11 18:38:00,2014-05-11 18:39:00,125.21978,Medium Gap,True,0
2,3.0,2010,10,Minnesota,MRO,-1.5,cold,severe weather,3000.0,NaN,...,5310903.0,73.27,2279.0,18.2,2010-10-26 20:00:00,2010-10-28 22:00:00,125.21978,Medium Gap,False,1
3,4.0,2012,6,Minnesota,MRO,-0.1,normal,severe weather,2550.0,NaN,...,5380443.0,73.27,2279.0,18.2,2012-06-19 04:30:00,2012-06-20 23:00:00,125.21978,Medium Gap,False,0
4,5.0,2015,7,Minnesota,MRO,1.2,warm,severe weather,1740.0,250.0,...,5489594.0,73.27,2279.0,18.2,2015-07-18 02:00:00,2015-07-19 07:00:00,125.21978,Medium Gap,False,0


In [245]:
# Check what columns you actually have
print(outage_clean.columns.tolist())

['OBS', 'YEAR', 'MONTH', 'U.S._STATE', 'NERC.REGION', 'ANOMALY.LEVEL', 'CLIMATE.CATEGORY', 'CAUSE.CATEGORY', 'OUTAGE.DURATION', 'DEMAND.LOSS.MW', 'CUSTOMERS.AFFECTED', 'TOTAL.PRICE', 'COM.PERCEN', 'TOTAL.CUSTOMERS', 'PC.REALGSP.STATE', 'POPULATION', 'POPPCT_URBAN', 'POPDEN_URBAN', 'POPDEN_RURAL', 'OUTAGE.START', 'OUTAGE.RESTORATION', 'DENSITY_GAP', 'DENSITY_TIER', 'CUST_MISSING', 'SEVERE']


In [246]:
# Quick look at the distribution
print(outage_clean['DENSITY_GAP'].describe())

count    1523.000000
mean      164.409679
std       234.928429
min        12.100492
25%        33.100437
50%        76.216117
75%       160.217105
max      2367.714286
Name: DENSITY_GAP, dtype: float64


In [247]:
# Which state is the extreme outlier? 
outage_clean.nlargest(15, 'DENSITY_GAP')[['U.S._STATE', 'POPDEN_URBAN', 'POPDEN_RURAL', 'DENSITY_GAP']]

,U.S._STATE,POPDEN_URBAN,POPDEN_RURAL,DENSITY_GAP
1484,Nevada,3314.8,1.4,2367.714286
1485,Nevada,3314.8,1.4,2367.714286
1486,Nevada,3314.8,1.4,2367.714286
1487,Nevada,3314.8,1.4,2367.714286
1488,Nevada,3314.8,1.4,2367.714286
1489,Nevada,3314.8,1.4,2367.714286
1490,Nevada,3314.8,1.4,2367.714286
1351,Wyoming,1876.2,2.0,938.100000
1352,Wyoming,1876.2,2.0,938.100000
1353,Wyoming,1876.2,2.0,938.100000


In [248]:
# Describe 'DENSITY_GAP'
print(outage_clean['DENSITY_GAP'].describe())

count    1523.000000
mean      164.409679
std       234.928429
min        12.100492
25%        33.100437
50%        76.216117
75%       160.217105
max      2367.714286
Name: DENSITY_GAP, dtype: float64


In [249]:
# Univariate Analysis 1
    # Analysing one variable(column) 'OUTAGE.DURATION'

fig_uni1 = px.histogram(
    outage_clean.dropna(subset=['OUTAGE.DURATION']),
    x='OUTAGE.DURATION',
    nbins=80,
    title='Distribution of Outage Duration',
    labels={'OUTAGE.DURATION': 'Duration (minutes)'},
    range_x=[0, 15000],
)
fig_uni1.update_layout(yaxis_title='Count', width=800, height=400)
fig_uni1.show()

# The histogram shows that most outages are relatively short, with the majority occurring under 2000 minutes. 
# However, there are a few extreme outages that last much longer, creating a right-skewed distribution.

In [250]:
# Univariate Analysis 2
# Cause category bar chart — always works and is more relevant
cause_counts = outage_clean['CAUSE.CATEGORY'].value_counts().reset_index()
cause_counts.columns = ['CAUSE.CATEGORY', 'COUNT']

fig_uni2 = px.bar(
    cause_counts,
    x='CAUSE.CATEGORY',
    y='COUNT',
    color='CAUSE.CATEGORY',
    title='Number of Outages by Cause Category',
    labels={'COUNT': 'Number of Outages', 'CAUSE.CATEGORY': 'Cause Category'},
)
fig_uni2.update_layout(showlegend=False, xaxis_tickangle=-30, width=800, height=400)
fig_uni2.show()

# The bar chart shows that severe weather accounts for the largest number of outages by a significant margin. 
# Intentional attacks are the second most frequent cause, while other categories occur far less often.

In [251]:
# Bivariate Analysis 1
fig_bi1 = px.box(
    outage_clean.dropna(subset=['OUTAGE.DURATION', 'DENSITY_TIER']),
    x='DENSITY_TIER',
    y='OUTAGE.DURATION',
    color='CAUSE.CATEGORY',
    log_y=True,
    category_orders={'DENSITY_TIER': [
        'Low Gap', 'Medium Gap', 'High Gap'
    ]},
    title='Outage Duration by Density Gap and Cause',
    labels={
        'OUTAGE.DURATION': 'Duration (minutes)',
        'DENSITY_TIER': 'Density Gap',
        'CAUSE.CATEGORY': 'Cause'
    },
)
fig_bi1.update_layout(width=900, height=500)
fig_bi1.show()

# This box plot shows the relationship between population density gap and outage duration across different outage causes.
# Outages in higher density-gap states tend to have slightly longer durations and more variability, especially for severe weather events.

In [252]:
# Bivariate Analysis 2
fig_bi2 = px.box(
    outage_clean.dropna(subset=['CUSTOMERS.AFFECTED', 'DENSITY_TIER']),
    x='DENSITY_TIER',
    y='CUSTOMERS.AFFECTED',
    color='DENSITY_TIER',
    log_y=True,
    category_orders={'DENSITY_TIER': [
        'Low Gap', 'Medium Gap', 'High Gap'
    ]},
    title='Customers Affected by Urban-Rural Density Gap',
    labels={
        'CUSTOMERS.AFFECTED': 'Customers Affected',
        'DENSITY_TIER': 'Density Gap'
    },
)
fig_bi2.update_layout(showlegend=False, width=750, height=450)
fig_bi2.show()

# This plot shows how the number of customers affected varies across density tiers.
# Low Gap states tend to affect more customers per outage (median ~100k),
# while High Gap states show a much wider spread with many smaller outages.

In [253]:
# Interesting Aggregates

pivot_duration = outage_clean.pivot_table(
    values='OUTAGE.DURATION',
    index='CAUSE.CATEGORY',
    columns='DENSITY_TIER',
    aggfunc='median'
)

pivot_duration

# This pivot table shows the median outage duration for each outage cause across population density tiers. 
# Severe weather outages consistently have the longest durations across all density tiers, 
# while other causes such as equipment failure and public appeals tend to have shorter outages. 
# The table helps show how outage duration varies depending on both the cause and the population density characteristics of a state.

/var/folders/1b/54c6c0t12kn6cl16xckz50w80000gn/T/ipykernel_4593/543790025.py:3: FutureWarning:

The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior



DENSITY_TIER,Low Gap,Medium Gap,High Gap
CAUSE.CATEGORY,,,
equipment failure,376.0,224.0,202.0
fuel supply emergency,7565.5,12240.0,1080.0
intentional attack,1.0,90.0,70.0
islanding,71.5,100.5,78.5
public appeal,1207.5,488.5,402.0
severe weather,2760.0,2625.5,1500.0
system operability disruption,163.5,155.0,227.0


## Step 3: Assessment of Missingness

In [254]:
# ASSESSMENT OF MISSINGNESS

# Check missingness rates
missing = outage_clean.isnull().sum()
missing_pct = (missing / len(outage_clean) * 100).round(1)
missing_pct[missing_pct > 0].sort_values(ascending=False).head(10)

DEMAND.LOSS.MW        46.0
CUSTOMERS.AFFECTED    28.9
OUTAGE.DURATION        3.7
OUTAGE.RESTORATION     3.7
TOTAL.PRICE            1.4
COM.PERCEN             1.4
POPDEN_RURAL           0.7
DENSITY_GAP            0.7
DENSITY_TIER           0.7
MONTH                  0.5
dtype: float64

### MNAR Analysis

We believe the column `DEMAND.LOSS.MW` (megawatts of electricity demand lost) is likely **NMAR** (Not Missing at Random). When a power outage causes very little demand loss, utilities may not bother measuring or reporting it — meaning the value is missing *because* the value itself is small. This makes the missingness depend on the actual (unobserved) value, which is the definition of MNAR.

To make this column MAR instead, we could collect additional data on **utility reporting policies** — for example, whether each utility has a minimum threshold for reporting demand loss. If missingness is explained by the reporting policy rather than the value itself, it would be MAR.

In [255]:
# Create missingness indicator 
print(f"Missing: {outage_clean['CUST_MISSING'].sum()}")
print(f"Not missing: {(~outage_clean['CUST_MISSING']).sum()}")

Missing: 443
Not missing: 1090


In [256]:
# Permutation test, DEPENDS on CAUSE.CATEGORY
def tvd(dist1, dist2):
    combined = pd.concat([dist1, dist2], axis=1).fillna(0)
    return (combined.iloc[:, 0] - combined.iloc[:, 1]).abs().sum() / 2

missing_dist = outage_clean[outage_clean['CUST_MISSING']]['CAUSE.CATEGORY'].value_counts(normalize=True)
not_missing_dist = outage_clean[~outage_clean['CUST_MISSING']]['CAUSE.CATEGORY'].value_counts(normalize=True)
observed_tvd = tvd(missing_dist, not_missing_dist)
print(f'Observed TVD: {observed_tvd:.4f}')

labels = outage_clean['CUST_MISSING'].values
causes = outage_clean['CAUSE.CATEGORY'].values

n_reps = 1000
tvds = []
for _ in range(n_reps):
    shuffled = np.random.permutation(labels)
    miss_dist = pd.Series(causes[shuffled]).value_counts(normalize=True)
    not_dist = pd.Series(causes[~shuffled]).value_counts(normalize=True)
    tvds.append(tvd(miss_dist, not_dist))

p_value_cause = np.mean(np.array(tvds) >= observed_tvd)
print(f'P-value: {p_value_cause}')

Observed TVD: 0.5581
P-value: 0.0


In [257]:
# Visualize Cause distribution when missing vs not
cause_missing = outage_clean.groupby(['CAUSE.CATEGORY', 'CUST_MISSING']).size().reset_index(name='count')
cause_missing['CUST_MISSING'] = cause_missing['CUST_MISSING'].map({True: 'Missing', False: 'Not Missing'})

totals = cause_missing.groupby('CUST_MISSING')['count'].transform('sum')
cause_missing['proportion'] = cause_missing['count'] / totals

fig_miss1 = px.bar(
    cause_missing,
    x='CAUSE.CATEGORY',
    y='proportion',
    color='CUST_MISSING',
    barmode='group',
    title=f'Distribution of Cause Category by CUSTOMERS.AFFECTED Missingness (TVD={observed_tvd:.3f}, p={p_value_cause})',
    labels={
        'proportion': 'Proportion',
        'CAUSE.CATEGORY': 'Cause Category',
        'CUST_MISSING': 'CUSTOMERS.AFFECTED'
    },
)
fig_miss1.update_layout(xaxis_tickangle=-30, width=850, height=450)
fig_miss1.show()

In [258]:
# Empirical distribution; Cause test
fig_perm1 = px.histogram(
    x=tvds,
    nbins=50,
    title=f'Permutation Test: Missingness vs CAUSE.CATEGORY (p={p_value_cause})',
    labels={'x': 'Total Variation Distance'},
    opacity=0.7,
)
fig_perm1.add_vline(x=observed_tvd, line_dash='dash', line_color='red',
                    annotation_text=f'Observed TVD: {observed_tvd:.3f}')
fig_perm1.update_layout(width=800, height=400, showlegend=False)
fig_perm1.show()

In [259]:
# Permutation test; does NOT depend on 'COM.PERCEN'
subset_com = outage_clean.dropna(subset=['COM.PERCEN']).copy()
labels_com = subset_com['CUST_MISSING'].values
com_vals = subset_com['COM.PERCEN'].values

obs_diff_com = abs(com_vals[labels_com].mean() - com_vals[~labels_com].mean())
print(f'Observed diff in mean COM.PERCEN: {obs_diff_com:.4f}')

n_reps = 1000
diffs_com = []
for _ in range(n_reps):
    shuffled = np.random.permutation(labels_com)
    diff = abs(com_vals[shuffled].mean() - com_vals[~shuffled].mean())
    diffs_com.append(diff)

p_value_com = np.mean(np.array(diffs_com) >= obs_diff_com)
print(f'P-value: {p_value_com}')

Observed diff in mean COM.PERCEN: 0.0085
P-value: 0.986


In [260]:
# Visualize 'COM.PERCEN' distribution when missing vs not
fig_miss_com = px.histogram(
    subset_com,
    x='COM.PERCEN',
    color='CUST_MISSING',
    barmode='overlay',
    opacity=0.6,
    nbins=40,
    color_discrete_map={True: 'red', False: 'steelblue'},
    title=f'Distribution of COM.PERCEN by CUSTOMERS.AFFECTED Missingness (p={p_value_com})',
    labels={
        'COM.PERCEN': 'Commercial Electricity Consumption (%)',
        'CUST_MISSING': 'CUSTOMERS.AFFECTED Missing'
    },
)
fig_miss_com.update_layout(width=800, height=400)
fig_miss_com.show()

In [261]:
# Empirical distribution 'COM.PERCEN' test 
fig_perm_com = px.histogram(
    x=diffs_com,
    nbins=50,
    title=f'Permutation Test: Missingness vs COM.PERCEN (p={p_value_com})',
    labels={'x': 'Absolute Difference in Mean COM.PERCEN'},
    opacity=0.7,
)
fig_perm_com.add_vline(x=obs_diff_com, line_dash='dash', line_color='red',
                       annotation_text=f'Observed: {obs_diff_com:.4f}')
fig_perm_com.update_layout(width=800, height=400, showlegend=False)
fig_perm_com.show()

# STEP 4: Hypothesis Testing


In [262]:
# STEP 4: HYPOTHESIS TESTING

# Question: Do states with a high urban-rural density gap have
# significantly shorter outages than states with a low density gap?

# H0: Mean duration is the same for high-gap and low-gap states
# H1: Low-gap states have higher mean duration than high-gap states

combined = outage_clean.dropna(subset=['OUTAGE.DURATION', 'DENSITY_TIER'])
combined = combined[combined['DENSITY_TIER'].isin(['High Gap', 'Low Gap'])].reset_index(drop=True)

durations = combined['OUTAGE.DURATION'].values
labels = combined['DENSITY_TIER'].values

observed_diff = (
    durations[labels == 'Low Gap'].mean() -
    durations[labels == 'High Gap'].mean()
)
print(f'Observed difference: {observed_diff:.2f} minutes ({observed_diff/60:.2f} hours)')

n_reps = 10000
diffs = []
for _ in range(n_reps):
    shuffled = np.random.permutation(labels)
    low_mean = durations[shuffled == 'Low Gap'].mean()
    high_mean = durations[shuffled == 'High Gap'].mean()
    diffs.append(low_mean - high_mean)

p_value = np.mean(np.array(diffs) >= observed_diff)
print(f'P-value: {p_value}')

Observed difference: 906.08 minutes (15.10 hours)
P-value: 0.0007


In [263]:
# Visualize
fig_hyp = px.histogram(
    x=diffs,
    nbins=50,
    title=f'Permutation Test: Low Gap vs High Gap Duration (p={p_value})',
    labels={'x': 'Difference in Mean Duration (minutes)'},
    opacity=0.7,
)
fig_hyp.add_vline(x=observed_diff, line_dash='dash', line_color='red',
                  annotation_text=f'Observed: {observed_diff:.0f} min')
fig_hyp.update_layout(width=800, height=400, showlegend=False)
fig_hyp.show()

### Conclusion

With a p-value of 0.0007, we reject the null hypothesis at a significance level of α = 0.05. There is strong evidence that Low Gap states (where population is more evenly spread) have significantly longer outage durations than High Gap states (where cities are much denser than rural areas). The observed difference is about 906 minutes (~15 hours) longer on average for Low Gap states.

## Step 5: Framing a Prediction Problem

In [264]:
# Prediction problem: Can we predict whether a power outage will be "severe"

# Goal: Predict if an outage will be "severe" (top 25% longest) or not.

# Type: Binary classification (1 = severe, 0 = not severe)

# Metric: F1 score, because 75% of outages are "not severe".
    # If we just guessed "not severe" every time we'd get 75% accuracy,
    # so accuracy is a bad metric here. F1 is better for imbalanced data.

# We only use features that would be available WHEN the outage starts.
    # We can't use duration, customers affected, or restoration time
    # because we wouldn't know those yet.

threshold = outage_clean['OUTAGE.DURATION'].quantile(0.75)
print(f'Severity threshold: {threshold:.0f} minutes ({threshold/60:.1f} hours)')
print(f'\nClass distribution:')
print(outage_clean['SEVERE'].value_counts())
print(f'\nBaseline accuracy (always predict not severe): {1 - outage_clean["SEVERE"].mean():.3f}')

Severity threshold: 2880 minutes (48.0 hours)

Class distribution:
SEVERE
0    1156
1     377
Name: count, dtype: Int64

Baseline accuracy (always predict not severe): 0.754


## Step 6: Baseline Model

### Baseline Model: Logistic Regression

**Features used:**
- `CAUSE.CATEGORY` — **nominal** (no natural order), one-hot encoded
- `NERC.REGION` — **nominal** (no natural order), one-hot encoded
- `ANOMALY.LEVEL` — **quantitative**, standardized with StandardScaler
- `MONTH` — **ordinal** (has natural order 1–12), standardized with StandardScaler

These 4 features are all known at the time an outage begins, so there is no data leakage. We chose Logistic Regression as the baseline because it is simple, fast, and gives us a reasonable starting point to compare against the final model.

In [265]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, f1_score

In [266]:
cat_features = ['CAUSE.CATEGORY', 'NERC.REGION']
num_features  = ['ANOMALY.LEVEL', 'MONTH']

model_df = outage_clean.dropna(subset=['SEVERE'] + cat_features + num_features).copy()
model_df['SEVERE'] = model_df['SEVERE'].astype(int)

X = model_df[cat_features + num_features]
y = model_df['SEVERE']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [267]:
preprocessor = ColumnTransformer(transformers=[
    ('cat', OneHotEncoder(handle_unknown='ignore', sparse_output=False), cat_features),
    ('num', StandardScaler(), num_features),
])

pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', LogisticRegression(max_iter=1000)),
])

pipeline.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('cat',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False),
                                                  ['CAUSE.CATEGORY',
                                                   'NERC.REGION']),
                                                 ('num', StandardScaler(),
                                                  ['ANOMALY.LEVEL',
                                                   'MONTH'])])),
                ('model', LogisticRegression(max_iter=1000))])

In [268]:
train_acc = accuracy_score(y_train, pipeline.predict(X_train))
test_acc  = accuracy_score(y_test,  pipeline.predict(X_test))
train_f1  = f1_score(y_train, pipeline.predict(X_train))
test_f1   = f1_score(y_test,  pipeline.predict(X_test))

print(f"Training Accuracy: {train_acc:.4f} | Training F1: {train_f1:.4f}")
print(f"Test Accuracy:     {test_acc:.4f} | Test F1:     {test_f1:.4f}")

Training Accuracy: 0.7869 | Training F1: 0.5752
Test Accuracy:     0.7803 | Test F1:     0.5379


## Step 7: Final Model

In [269]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score, f1_score

**Hyperparameters to tune:**
- **n_estimators:** more trees = more stable predictions, but diminishing returns
- **max_depth:** controls overfitting; too deep = memorizes training data
- **min_samples_split:** prevents splits on too little data, helps generalization
- **max_features:** how many features each tree considers; reduces correlation between trees

In [270]:
cat_features = ['CAUSE.CATEGORY', 'NERC.REGION', 'DENSITY_TIER']
num_features  = ['ANOMALY.LEVEL', 'MONTH', 'TOTAL.CUSTOMERS']

model_df = outage_clean.dropna(subset=['SEVERE'] + cat_features + num_features).copy()
model_df['SEVERE'] = model_df['SEVERE'].astype(int)

X = model_df[cat_features + num_features]
y = model_df['SEVERE']

# Use same train/test split as baseline
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [271]:
preprocessor = ColumnTransformer(transformers=[
    ('cat', OneHotEncoder(handle_unknown='ignore', sparse_output=False), cat_features),
    ('num', StandardScaler(), num_features),
])

pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', RandomForestClassifier(random_state=42)),
])

param_grid = {
    'model__n_estimators':     [50, 100, 200],
    'model__max_depth':        [5, 10, None],
    'model__min_samples_split':[2, 5, 10],
    'model__max_features':     ['sqrt', 'log2'],
}

In [272]:
grid_search = GridSearchCV(
    pipeline, param_grid, cv=5, scoring='f1', n_jobs=-1, verbose=1
)
grid_search.fit(X_train, y_train)

print("Best hyperparameters:", grid_search.best_params_)
print()

# Refit on full training set with best params
best_model = grid_search.best_estimator_

Fitting 5 folds for each of 54 candidates, totalling 270 fits
Best hyperparameters: {'model__max_depth': 10, 'model__max_features': 'sqrt', 'model__min_samples_split': 2, 'model__n_estimators': 50}



In [273]:
train_acc = accuracy_score(y_train, best_model.predict(X_train))
test_acc  = accuracy_score(y_test,  best_model.predict(X_test))
train_f1  = f1_score(y_train, best_model.predict(X_train))
test_f1   = f1_score(y_test,  best_model.predict(X_test))

print(f"Training Accuracy: {train_acc:.4f} | Training F1: {train_f1:.4f}")
print(f"Test Accuracy:     {test_acc:.4f} | Test F1:     {test_f1:.4f}")

Training Accuracy: 0.9208 | Training F1: 0.8400
Test Accuracy:     0.8317 | Test F1:     0.5920


In [274]:
from sklearn.metrics import classification_report

y_pred = best_model.predict(X_test)
print(classification_report(y_test, y_pred, target_names=['Not Severe', 'Severe']))

              precision    recall  f1-score   support

  Not Severe       0.88      0.91      0.89       236
      Severe       0.64      0.55      0.59        67

    accuracy                           0.83       303
   macro avg       0.76      0.73      0.74       303
weighted avg       0.82      0.83      0.83       303



## Step 8: Fairness Analysis

**Question:** Does our model perform worse for outages in "High Gap" states (large urban-rural density difference) compared to "Low Gap" states (more evenly spread population)?

- **Group X:** Low Gap (low urban-rural density difference)
- **Group Y:** High Gap (high urban-rural density difference)
- **Evaluation Metric:** Precision of the "Severe" class — because our model is a classifier, precision tells us what fraction of predicted-severe outages are actually severe.
- **Null Hypothesis:** The model is fair. Its precision for Low Gap states and High Gap states is roughly the same, and any difference is due to random chance.
- **Alternative Hypothesis:** The model is unfair. Its precision for Low Gap states is lower than its precision for High Gap states.
- **Test Statistic:** precision(High Gap) − precision(Low Gap). A positive value means the model is more precise for High Gap.
- **Significance Level:** 0.05

In [275]:
from sklearn.metrics import precision_score

test_df = X_test.copy()
test_df['SEVERE_TRUE'] = y_test.values
test_df['SEVERE_PRED'] = best_model.predict(X_test)

low_gap  = test_df[test_df['DENSITY_TIER'] == 'Low Gap']
high_gap = test_df[test_df['DENSITY_TIER'] == 'High Gap']

prec_low  = precision_score(low_gap['SEVERE_TRUE'],  low_gap['SEVERE_PRED'],  zero_division=0)
prec_high = precision_score(high_gap['SEVERE_TRUE'], high_gap['SEVERE_PRED'], zero_division=0)

observed_diff = prec_high - prec_low

print(f"Precision (Low Gap):  {prec_low:.4f}  (n={len(low_gap)})")
print(f"Precision (High Gap): {prec_high:.4f}  (n={len(high_gap)})")
print(f"Observed difference (High − Low): {observed_diff:.4f}")



Precision (Low Gap):  0.6552  (n=102)
Precision (High Gap): 0.6667  (n=103)
Observed difference (High − Low): 0.0115


In [276]:
# Permutation Test
np.random.seed(42)
n_permutations = 1000

combined = pd.concat([low_gap, high_gap])
n_high = len(high_gap)

perm_diffs = []
for _ in range(n_permutations):
    shuffled = combined.sample(frac=1, replace=False)
    perm_high = shuffled.iloc[:n_high]
    perm_low  = shuffled.iloc[n_high:]
    
    p_high = precision_score(perm_high['SEVERE_TRUE'], perm_high['SEVERE_PRED'], zero_division=0)
    p_low  = precision_score(perm_low['SEVERE_TRUE'],  perm_low['SEVERE_PRED'],  zero_division=0)
    perm_diffs.append(p_high - p_low)

perm_diffs = np.array(perm_diffs)

p_value_fair = np.mean(perm_diffs >= observed_diff)
print(f"\np-value: {p_value_fair:.4f}")

if p_value_fair < 0.05:
    print("Result: Reject the null hypothesis at α = 0.05.")
    print("The model appears UNFAIR — precision differs significantly between groups.")
else:
    print("Result: Fail to reject the null hypothesis at α = 0.05.")
    print("No significant evidence of unfairness — precision difference is consistent with random chance.")


p-value: 0.4870
Result: Fail to reject the null hypothesis at α = 0.05.
No significant evidence of unfairness — precision difference is consistent with random chance.


In [277]:
fig_fair = px.histogram(
    x=perm_diffs,
    nbins=50,
    title=f'Fairness Permutation Test: Precision Difference (p={p_value_fair:.4f})',
    labels={'x': 'Precision(High Gap) − Precision(Low Gap)', 'y': 'Count'},
    opacity=0.7,
)

fig_fair.add_vline(
    x=observed_diff,
    line_dash='dash',
    line_color='red',
    annotation_text=f'Observed diff = {observed_diff:.4f}',
    annotation_position='top right',
)

fig_fair.update_layout(
    xaxis_title='Precision(High Gap) − Precision(Low Gap)',
    yaxis_title='Count',
    showlegend=False,
)

fig_fair.show()

In [278]:
# Create assets folder
import os
os.makedirs('assets', exist_ok=True)

# Step 2 plots
fig_uni1.write_html('assets/fig_uni1.html', include_plotlyjs='cdn')
fig_uni2.write_html('assets/fig_uni2.html', include_plotlyjs='cdn')
fig_bi1.write_html('assets/fig_bi1.html', include_plotlyjs='cdn')

# Step 3 plots
fig_perm1.write_html('assets/fig_perm1.html', include_plotlyjs='cdn')
fig_perm_com.write_html('assets/fig_perm_com.html', include_plotlyjs='cdn')

# Step 4 plot
fig_hyp.write_html('assets/fig_hyp.html', include_plotlyjs='cdn')

# Step 8 plot
fig_fair.write_html('assets/fig_fair.html', include_plotlyjs='cdn')